moveit_ros/planning/moveit_cpp/src/moveit_cpp.cpp

```diff
@@ -238,14 +238,54 @@ bool MoveItCpp::execute(const std::string& group_name, const robot_trajectory::R
   // Execute trajectory
   moveit_msgs::RobotTrajectory robot_trajectory_msg;
   robot_trajectory->getRobotTrajectoryMsg(robot_trajectory_msg);
-  // TODO: cambel
-  // blocking is the only valid option right now. Add non-bloking use case
+  moveit_controller_manager::ExecutionStatus execution_status;
+
+  static std::mutex execution_complete_mutex_;
+  static std::condition_variable execution_complete_condition_;
+  auto callback = trajectory_execution_manager::TrajectoryExecutionManager::ExecutionCompleteCallback();
+
+  static int trajectory_id = 0;
+  trajectory_id++;
+
+  if (trajectory_execution_manager_->getEnableSimultaneousExecution())
+  {
+    callback = [this,
+                &execution_status](__attribute__((unused)) const moveit_controller_manager::ExecutionStatus status) {
+      {
+        std::unique_lock<std::mutex> ulock(execution_complete_mutex_);
+        std::remove(active_trajectories_.begin(), active_trajectories_.end(), trajectory_id);
+      }
+      execution_complete_condition_.notify_all();
+      execution_status = status;
+    };
+  }
+
+  if (!trajectory_execution_manager_->push(robot_trajectory_msg, "", callback))
+    return false;
+
+  {
+    std::unique_lock<std::mutex> ulock(execution_complete_mutex_);
+    active_trajectories_.push_back(trajectory_id);
+  }
+
   if (blocking)
```

Comment:
@cambel there is no case for `blocking==false`, right? We were thinking about removing this blocking flag completely in MoveIt2.

Reply:
Hi @AndyZe
For this new feature, `blocking==false` is used to send multiple trajectories to be simultaneously executed without waiting for each one to finish before executing the next one, as in this demo: 

https://github.com/cambel/moveit_simultaneous_motions_demo/blob/2c9b7f6077e59cea73b544107b7bda7e6fbb8ca1/panda_dual_arm_demo/src/simple_path_moveitcpp.cpp#L168-L170

Reply:
OK, thanks for the reply.

Shouldn't there be an `else()` here? Like this:

if (blocking)
{
  ...
}
else
{
  // Still need to execute the trajectory somehow, right?
}

Reply:
Thank you, I was missing the case where simultaneous execution mode is OFF and `blocking==false`.